### ISSUE TITLE
Allow protocol discovery between client and server.

### ISSUE THREAD

**Issue by timemachine3030** [tus/tus-resumable-upload-protocol#29]
If writing an application that is able to upload to many different servers, the application will need to check if the remote server supports this protocol. Therefore, I suggest the following protocol extension to start a conversion on the topic.

(Note: that servers that do not support the tus protocol will reply to `HEAD` requests with either `200 OK` or `404 Not Found`)

### Protocol Discovery

All implementations SHOULD include the protocol discovery API to allow clients to negotiate, with a server, a common supported version of this protocol.

When making a `HEAD` request for a new file the client MAY include a `TUS-Resumable` header with the version of this protocol. 

When receiving a `HEAD` request with the `TUS-Resumable` header the server MAY reply with the `202 Accepted` status and the current `Offset`.

#### Example

**Request:**

```
HEAD /files/24e533e02ec3bc40c387f1a0e460e216 HTTP/1.1
Host: tus.example.org
TUS-Resumable: 1.0
```

**Response:**

```
HTTP/1.1 202 Accepted
Offset: 0
```

**Comment by Acconut** [comment]
See #48 for the proposed changes.

**Comment by qsorix** [comment]
Sorry for a late response, I stayed offline duringe christmas.

@Acconut, please remind me why for TUS-Version only server is mentioned? Clients are expected to first send OPTIONS, then pick the version and send requests using the proper one? If so, I'm +1.

Will it say anywhere what happens when server doesn't like client's version? I.e. what's the proper way to reject invalid requests?

I think it would help if every header was accompanied by a sensible example, to limit possible confusion regarding formating etc.

**Comment by Acconut** [comment]
> Sorry for a late response, I stayed offline duringe christmas.

No problem, we still have time until 1.0 needs to be done. :)

> I think it would help if every header was accompanied by a sensible example, to limit possible confusion regarding formating etc.

I agree. An example for an `OPTIONS` request and response including all headers will follow.

> @Acconut, please remind me why for TUS-Version only server is mentioned? Clients are expected to first send OPTIONS, then pick the version and send requests using the proper one? If so, I'm +1.

Yes, that's how it is intended by me. The server doesn't need to know which versions the client supports, it must only be aware of the version the client is using in its requests.

> Will it say anywhere what happens when server doesn't like client's version? I.e. what's the proper way to reject invalid requests?

Good point I forgot about! I vote for `412 Precondition Failed`. It is intended to be used when the server can't process the request since the client specified conditions the server can't meet. And that's exactly what we need when the requester uses a version which is not supported by the responder.

**Comment by qsorix** [comment]
I'm fine with `412`.

**Comment by Acconut** [comment]
See a7ba01aa057c5ea172752a16b97a4e9a3d0b9470.

**Comment by qsorix** [comment]
+1

**Comment by Acconut** [comment]
Merged in 7c4fd4be02dd46628860b39d43a837ec57f80ced.
